src/components/views/voip/VideoFeed.tsx

```diff
@@ -18,67 +18,154 @@ import classnames from 'classnames';
+            // load() explicitly, it shouldn't be a problem. - Dave
+            currentMedia.play()
+        } catch (e) {
+            logger.info("Failed to play media element with feed", this.props.feed, e);
         }
     }
 
     componentWillUnmount() {
-        this.vid.current.removeEventListener('resize', this.onResize);
+        this.props.feed.removeListener(CallFeedEvent.NewStream, this.onNewStream);
+        this.video.current?.removeEventListener('resize', this.onResize);
+
+        const currentMedia = this.getCurrentMedia();
+        currentMedia.pause();
+        currentMedia.srcObject = null;
+        // As per comment in componentDidMount, setting the sink ID back to the
+        // default once the call is over makes setSinkId work reliably. - Dave
+        // Since we are not using the same element anymore, the above doesn't
+        // seem to be necessary - Šimon
     }
 
-    private setVideoElement() {
-        if (this.props.type === VideoFeedType.Local) {
-            this.props.call.setLocalVideoElement(this.vid.current);
-        } else {
-            this.props.call.setRemoteVideoElement(this.vid.current);
-        }
+    private getCurrentMedia() {
+        return this.audio.current || this.video.current;
     }
 
-    onResize = (e) => {
-        if (this.props.onResize) {
+    private onNewStream = () => {
+        this.setState({
+            audioMuted: this.props.feed.isAudioMuted(),
+            videoMuted: this.props.feed.isVideoMuted(),
+        });
+        this.playMedia();
+    };
+
+    private onResize = (e) => {
+        if (this.props.onResize && !this.props.feed.isLocal()) {
             this.props.onResize(e);
         }
     };
 
     render() {
         const videoClasses = {
             mx_VideoFeed: true,
-            mx_VideoFeed_local: this.props.type === VideoFeedType.Local,
-            mx_VideoFeed_remote: this.props.type === VideoFeedType.Remote,
+            mx_VideoFeed_local: this.props.feed.isLocal(),
+            mx_VideoFeed_remote: !this.props.feed.isLocal(),
+            mx_VideoFeed_voice: this.state.videoMuted,
+            mx_VideoFeed_video: !this.state.videoMuted,
             mx_VideoFeed_mirror: (
-                this.props.type === VideoFeedType.Local &&
+                this.props.feed.isLocal() &&
                 SettingsStore.getValue('VideoView.flipVideoHorizontally')
             ),
         };
 
-        return <video className={classnames(videoClasses)} ref={this.vid} />;
+        if (this.state.videoMuted) {
+            const member = this.props.feed.getMember();
+            const avatarSize = this.props.pipMode ? 76 : 160;
+
+            return (
+                <div className={classnames(videoClasses)} >
+                    <MemberAvatar
+                        member={member}
+                        height={avatarSize}
+                        width={avatarSize}
+                    />
+                    <audio ref={this.audio}></audio>
```

Comment:
Thanks for the explanation on the audio / video feeds: I see where these have gone now, ie. they're always in the VideoFeed. I guess this is fine now since we always display either the large view or the PiP view, but does switching the remote audio between different audio elements cause a gap in playback? We used a single audio element that didn't get unmounted / remounted to avoid this kind of problem.

Reply:
Hmm, good point. I _think_ this will be fine now since this only happens when we get put on hold but might be a problem in future. It shouldn't be too hard to rework this though

Reply:
Switching in & out of the room the call is in was the case I was thinking about (since that will now cause the audio/video element to be unmounted and a new one mounted and hooked up to the stream.

Reply:
Hmmm, that's true. But I've just tested it and I can't hear any gap :thinking:

Reply:
We probably should put the audio element somewhere else (better safe than sorry) and we probably also will have to do it anyway at some point. Where should we place them though? I am thinking `MatrixChat.tsx` or `LoggedInView.tsx` but there might be a better place.

Reply:
Yeah, I think LoggedInView is fine, or we could also just have CallHandler append them to the DOM as necessary.

I can hear the gap here, fwiw: 

https://user-images.githubusercontent.com/986903/116107956-617afa80-a6ab-11eb-8a7c-81aa00b706e4.mov

Reply:
Reworked now - I've made the `CallHandler` into an `EventEmitter` so that it can emit when calls change. The `LoggedInView` then listens to these changes and renders `AudioFeedArrayForCall`s. `AudioFeedArrayForCall` is a group of `AudioFeeds` for a specific call. 

I hope this design makes sense, if not feel free to scream. Also, sorry for not thinking of this earlier